parsers/ZA.py

```diff
@@ -0,0 +1,965 @@
+                            + value["cleansed_csv"][comma_idx + starting_idx + 1 :]
+                        )
+
+                        # Update curr_update_dict from 2. CLEANING STEP
+                        left_from_period = value["cleansed_csv"][
+                            : comma_idx + starting_idx
+                        ].split(",")[-1]
+                        right_from_period = value["cleansed_csv"][
+                            comma_idx + starting_idx + 1 :
+                        ].split(",")[0]
+                        old_updated_value = left_from_period + "," + right_from_period
+                        new_updated_value = left_from_period + "." + right_from_period
+                        curr_update_dict[old_updated_value] = new_updated_value
+
+        curr_cleansed_csv_len = len(value["cleansed_csv"].split(","))
+
+        # 3. CLEANING STEP: if last_cleansed_csv_len != curr_cleansed_csv_len, then an update happened in the last few timestamps and value stayed the same until now. Go through cleansed_csv_value_updates in reverse order and check if there are matches with value['cleansed_csv']. If so, update value['cleansed_csv'] with the value from cleansed_csv_value_updates.
+
+        if last_cleansed_csv_len != curr_cleansed_csv_len:
+            for update_i in range(len(cleansed_csv_value_updates) - 1, -1, -1):
+                break_outer_loop = False
+                for key in cleansed_csv_value_updates[update_i]:
+                    if key in value["cleansed_csv"]:
+                        value["cleansed_csv"] = value["cleansed_csv"].replace(
+                            key, cleansed_csv_value_updates[update_i][key]
+                        )
+                        # Update curr_update_dict from 3. CLEANING STEP
+                        curr_update_dict[key] = cleansed_csv_value_updates[update_i][
+                            key
+                        ]
+                        break_outer_loop = True
+                        break
+                if break_outer_loop:
+                    break
+
+        curr_cleansed_csv_len = len(value["cleansed_csv"].split(","))
+
+        # This is useful if debugging corrupt rows
+        # ---------------------------------------
+        # if last_cleansed_csv_len != curr_cleansed_csv_len:
+        #     print(f"{i}/{len(merged_csv_bi_data)}: {key}")
+        #     print(f"last: {last_cleansed_csv_len} curr: {curr_cleansed_csv_len}")
+        #     pp.pprint(value)
+        #     print("-----------------------------------------------------")
+
+        last_cleansed_csv_len = curr_cleansed_csv_len
+        cleansed_csv_value_updates.append(curr_update_dict)
+
+    # Mapping columns to keys
+    # Helpful: https://www.eskom.co.za/dataportal/glossary/
+    column_mapping = {
+        0: "coal",  # Thermal_Gen_Excl_Pumping_and_SCO
+        1: "unknown",  # Eskom_OCGT_SCO_Pumping             changed to unknown since negative oil is not possible
```

Comment:
Suggestion: perhaps you could replace "unknown" with "ignored" to make it more intuitive these fields aren't being used? Except `19` of course, as that one is actually being used.

Reply:
Done in commit 024737e2f1e4c96773dd810f06a60fd479b69ecb